src/Planner/findParallelReplicasQuery.cpp

```diff
@@ -0,0 +1,431 @@
+            /// We can execute parallel replicas over storage::read.
+            if (!res)
+                return nullptr;
+
+            return subquery_node;
+        }
+
+        /// Query is simple enough to be fully distributed.
+        res = subquery_node;
+        prev_checked_node = curr_node;
+    }
+
+    return res;
+}
+
+const QueryNode * findParallelReplicasQuery(const QueryTreeNodePtr & query_tree_node, SelectQueryOptions & select_query_options)
+{
+    if (select_query_options.only_analyze)
+        return nullptr;
+
+    auto * query_node = query_tree_node->as<QueryNode>();
+    auto * union_node = query_tree_node->as<UnionNode>();
+
+    if (!query_node && !union_node)
+        throw Exception(ErrorCodes::UNSUPPORTED_METHOD,
+            "Expected QUERY or UNION node. Actual {}",
+            query_tree_node->formatASTForErrorMessage());
+
+    auto context = query_node ? query_node->getContext() : union_node->getContext();
+
+    if (!context->canUseParallelReplicasOnInitiator())
+        return nullptr;
+
+    auto stack = getSupportingParallelReplicasQuery(query_tree_node.get());
+    /// Empty stack means that storage does not support parallel replicas.
+    if (stack.empty())
+        return nullptr;
+
+    /// We don't have any subquery and storage can process parallel replicas by itself.
+    if (stack.top() == query_tree_node.get())
+        return nullptr;
+
+    /// This is needed to avoid infinite recursion.
+    auto mutable_context = Context::createCopy(context);
+    mutable_context->setSetting("allow_experimental_parallel_reading_from_replicas", Field(0));
+
+    /// Here we replace tables to dummy, in order to build a temporary query plan for parallel replicas analysis.
+    ResultReplacementMap replacement_map;
+    auto updated_query_tree = replaceTablesWithDummyTables(query_tree_node, mutable_context);
+
+    SelectQueryOptions options;
+    Planner planner(updated_query_tree, options, std::make_shared<GlobalPlannerContext>(nullptr, nullptr));
+    planner.buildQueryPlanIfNeeded();
+
+    /// This part is a bit clumsy.
+    /// We updated a query_tree with dummy storages, and mapping is using updated_query_tree now.
+    /// But QueryNode result should be taken from initial query tree.
+    /// So that we build a list of candidates again, and call findParallelReplicasQuery for it.
+    auto new_stack = getSupportingParallelReplicasQuery(updated_query_tree.get());
+    const auto & mapping = planner.getQueryNodeToPlanStepMapping();
+    const auto * res = findParallelReplicasQuery(new_stack, mapping);
+
+    /// Now, return a query from initial stack.
+    if (res)
+    {
+        while (!new_stack.empty())
+        {
+            if (res == new_stack.top())
+                return stack.top();
+
+            stack.pop();
+            new_stack.pop();
+        }
+    }
+
+    return res;
+}
+
+static const TableNode * findTableForParallelReplicas(const IQueryTreeNode * query_tree_node)
```

Comment:
I'd prefer to have a non recursive implementation.

Reply:
Agree.
However, it would make sense if we rewrite all the visitors to non-recursive mode implementation :)